doc/source/ray-air/check-ingest.rst

```diff
@@ -83,20 +87,31 @@ Enabling Streaming Ingest
 
     You should use bulk ingest when:
 
-     * you have enough memory to fit data blocks in cluster object store;
-     * your preprocessing step is expensive per each epoch; and
-     * you want best performance when both or either the above conditions are met.
+    * you have enough memory to fit data blocks in cluster object store; or
+    * your preprocessing step is expensive to recompute on each epoch
 
 .. tabbed:: Streaming Ingest (experimental)
 
-    In streaming ingest mode, :py:func:`~ray.air.session.get_dataset_shard` returns a :py:class:`~ray.data.dataset_pipeline.DatasetPipeline` pipeline that
-    can be used to read data in a streaming way.
-    To enable streaming ingest, set ``use_stream_api=True`` in the dataset config.
+    In streaming ingest mode, instead of loading the entire dataset into the
+    Ray object store at once, AIR will load a fraction of the dataset at a
+    time. This can be desirable when the dataset is very large, and caching it
+    all at once would cause expensive disk spilling. The downside is that the
+    dataset will have to be recomputed on each epoch, which can add overhead if
+    the computation is expensive.
+
+    To enable this mode, use the :py:meth:`max_object_store_memory_fraction
+    <ray.air.config.DatasetConfig>` argument. This argument defaults to -1,
+    meaning that bulk ingest should be used and the entire dataset should be
+    computed and cached before execution.
+
+    Use a float value between 0 and 1 to indicate the "window" size, i.e. the
+    maximum fraction of object store memory that should be used at once. A
+    reasonable value is 0.2, meaning 20% of available object store memory.
+    Larger window sizes can improve performance by increasing parallelism.
 
-    By default, this will tell AIR to load *windows* of 1GiB of data into memory at a time.
-    Performance can be increased with larger window sizes, which can be adjusted using the
-    ``stream_window_size`` config.
-    A reasonable stream window size is something like 20% of available object store memory.
+    The value 1 means that the window size is the full dataset. In other words,
+    Ray will compute a full copy of the dataset in parallel with the current
+    copy that is being iterated upon.
```

Comment:
This setting is just for compatibility with the previous behavior for global_shuffle=True, use_window_api=True. Once we have native pipelined Datasets execution, we won't need this setting anymore because users wouldn't need to set a DatasetPipeline window size.

> 1 is always strictly worse than -1 right, since the preprocessed dataset is not cached?

Not necessarily. -1 would cache the preprocessed dataset and have to do the global shuffle synchronously; 1 might use more memory but can be better for performance because it can overlap the shuffle.

Reply:
Yes, that's right. I realize the behavior for 1 is not very intuitive but I mainly did this to make sure that the behavior doesn't change from before. The problem is that the current DatasetPipeline implementation takes two different paths for windowing on/off, so I tried to keep this behavior by doing:
1. 0 < max_object_store_memory_fraction < 1 -> window
2. max_object_store_memory_fraction = 1 -> do not window

Let me check again; we might able to just remove the second codepath now.

Reply:
Okay, I was able to simplify the behavior here. Before:
1. default -> cache the preprocessed Dataset
2. default, global_shuffle=True -> cache the preprocessed Dataset and shuffle once
3. use_stream_api=True, stream_window_size=-1 -> cache the preprocessed Dataset (so basically same as 1 but user gets a DatasetPipeline instead of a Dataset)
4. use_stream_api=True, stream_window_size=-1, global_shuffle=True -> cache the preprocessed Dataset and shuffle each epoch
5. use_stream_api=True, stream_window_size=N -> reapply preprocessing on each epoch, with window size N

Now:
1. default -> cache the preprocessed Dataset
2. default, global_shuffle=True -> cache the preprocessed Dataset and shuffle each epoch (by converting to DatasetPipeline)
3. max_object_store_memory_fraction=F -> reapply preprocessing on each epoch, with window size (F * obj store memory)

Does this sound good? If so, will add this to the PR description too. Thanks for pushing on this, it seems much nicer now!

Reply:
Yes, this sgtm!